src/plugin/helpers/keymap.ts

```diff
@@ -26,3 +27,60 @@ export const createHistoryCommands = (
     return view?.dispatch(view.state.tr.scrollIntoView()) ?? false;
   },
 });
+
+/*
+  These commands restrict caret traversal to the current text block
+  When navigating with the arrow keys the caret should stop at the beginning and end of text blocks
+*/
+export const preventCaretBoundaryTraversalKeymap: Record<string, Command> = {
+  ArrowRight: (state: EditorState) => {
+    if (
+      state.selection.$from.parentOffset ===
+      state.selection.$from.parent.content.size
+    ) {
+      return true;
+    }
+
+    return false;
+  },
+  ArrowLeft: (state: EditorState) => {
+    if (state.selection.$from.parentOffset === 0) {
+      return true;
+    }
+
+    return false;
+  },
+  ArrowUp: (
+    state: EditorState,
+    dispatch?: (tr: Transaction) => void,
+    view?: EditorView
+  ) => {
+    if (view?.endOfTextblock("up", state)) {
+      // Move the cursor to the start of this block
+      const tr = state.tr.setSelection(
+        TextSelection.between(state.doc.resolve(0), state.doc.resolve(0))
```

Comment:
This might be a little clearer as:

```suggestion
        TextSelection.atStart(state.doc)
```

(there's `atEnd` for the other case, too)